docs: Add VM latency checkup design document (#12)
kiagnose/kiagnose: docs/vm_latency_checkup_design.md

# Summary

Provide the functionality to validate connectivity and measure latency on a network, between two virtual machines that
run on a [Kubernetes](https://kubernetes.io/) cluster with [KubeVirt](https://kubevirt.io/).

# Motivation

Cluster administrators are often required to check network connectivity and its performance on various networks. Such
checks may occur immediately after the cluster is deployed, as an acceptance test, when new networks are introduced or
just as part of a troubleshooting effort.

Whatever the reason may be, automating a basic connectivity test with a latency measurement can greatly improve the
maintainability of a cluster, reduce the load from the administrator, minimize potential mistakes from such a
multi-step operation and generally save time.

## Goals

- Validate connectivity between two VMI/s, on a predefined network.
- Measure latency between two VMI/s, on a predefined network.
- Share results (and logs) with the cluster administrator.

## Non-Goals

- Define or configure the network under test.

# Proposal

## Definition of Users

- Kubernetes Cluster Administrators

## User Stories

As a Kubernetes Cluster Administrator I would like to:

> 1. Check the connectivity and latency of a network in my cluster, by:

Reviewer: 1.1 and 1.2 seem to be requirements and not user stories, is it on purpose?
I think 1.2 and 1.3  can be combined, 1.3 seems redundant.

Author: This document is a 1 to 1 conversion of the original google doc file to markdown.
I suggest to look there for the reasoning.

Reviewer: The original not only this doc is not opened for anyone so referring to it for reasoning is not effecting.
Other than that, the reasoning for what I ask is not mentioned there.

>    1.1. Having an active cluster, with at least two worker nodes.

Reviewer: ```suggestion
   1.1. Having an active cluster, with at least one worker node.
```
Having more then one worker is not a must, one may want to check connectivity between two VMs on the same node.

Reviewer 2: I do not think so.
This is a requirement placed here explicitly and if the checkup is not making sure this is the case (e.g. use anti-affinity) then we have a problem/bug. If a user will specify the same node explicitly, I can accept it, but it should be an exception and not the common/recommended case.

>    1.2. Defining, configuring and creating a NetworkAttachmentDefinition, which I am interested in checking.

Reviewer: ```suggestion
   1.2. Create a NetworkAttachmentDefinition object of the network I am interested in checking.
```

>    1.3. Running the Kubevirt Network Latency checkup, while providing as input the NetworkAttachmentDefinition.

Reviewer: ```suggestion
   1.3. Running the Kubevirt Network Latency checkup, while providing the `NetworkAttachmentDefinition` as input.
```

Author: Done.

  spec.clusterRoles: |
     kubevirt-vm-latency-checker
  spec.param.network_attachment_definition_namespace: default
  spec.param.network_attachment_definition_name: br10
  spec.param.source_node: node01
  spec.param.target_node: node02
  spec.param.sample_duration_seconds: 300
  spec.param.max_desired_latency_milliseconds: 50
  status.succeeded: “true”
  status.failureReason: ””
  status.result.minLatencyNanoSec: 2000
  status.result.maxLatencyNanoSec: 3000
  status.result.avgLatencyNanoSec: 2500
  status.results.measurementDurationSec: 300
  status.result.sourceNode: node01
  status.result.targetNode: node02
```

## Component Dependency

The solution assumes the following components are available (and correctly set):

- Kubevirt is deployed and active in the cluster.
- [Multus](https://github.com/k8snetworkplumbingwg/multus-cni) deployed and ready.
- The [Network Attachment Definition](https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#heading=h.hylsbqoj5fxd) is defined and correctly configured, with a working backend network.

# Design Details

## Diagram
![block diagram](./images/latency_checkup_block_diagram.svg)

## Deliverables

> According to Kiagnose requirements.

Reviewer: This seems not effective, it should mention at least a container-image and documentation.
Alternatively this section can be added after Kiagnose docs are merged and add a ref.

Author: Added a link to the requirements.

  status.result.avgLatencyNanoSec: 2500
  status.results.measurementDurationSec: 300
  status.result.sourceNode: node01
  status.result.targetNode: node02
```

## Component Dependency

The solution assumes the following components are available (and correctly set):

- Kubevirt is deployed and active in the cluster.
- [Multus](https://github.com/k8snetworkplumbingwg/multus-cni) deployed and ready.
- The [Network Attachment Definition](https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#heading=h.hylsbqoj5fxd) is defined and correctly configured, with a working backend network.

# Design Details

## Diagram
![block diagram](./images/latency_checkup_block_diagram.svg)

## Deliverables

According to Kiagnose requirements.

## Main Processes

### Input Validation

The checkup shall verify all required inputs are present and valid.

### Prerequisites Validation

The checkup shall perform the following checks:

1. Check if Kubevirt is installed and ready.
2. Check if Multus exists.
3. Check if the input NetworkAttachmentDefinition exists.
4. There are at least two worker nodes.

### Setup

The checkup shall perform the following:

> 1. Check if the given NetworkAttachmentDefinition makes use of the SR-IOV CNI plugin.Create two VMI’s, each with node affinity according to the

Reviewer 3: didn't we change it so that we can wither use SRIOV or linux-bridge cni? If so, then we need to be accurate here.

Author: @RamLavi thanks for catching that.
I've changed the wording a little bit to make it more agnostic to CNI type.

  spec.roles: |
    <namespace/role name>
    <namespace/role name>
  status.succeeded: “true”
  status.failureReason: ””
  status.result.minLatencyNanoSec: 2000
  status.result.maxLatencyNanoSec: 3000
  status.result.avgLatencyNanoSec: 2500
  status.results.measurementDurationSec: 300
```

## Component Dependency

The solution assumes the following components are available (and correctly set):

- Kubevirt is deployed and active in the cluster.
- The [Network Attachment Definition](https://docs.google.com/document/d/1Ny03h6IDVy_e_vmElOqR7UdTPAG_RNydhVE1Kx54kFQ/edit#heading=h.hylsbqoj5fxd) is defined and correctly configured, with a working backend network.
- Multus deployed and ready.

# Design Details

## Diagram
![block diagram](./images/latency_checkup_block_diagram.png)

## Deliverables

According to the Frameworks requirements.

## Main Processes

### Input Validation

The checkup shall verify all required inputs are present and valid.

### Prerequisites Validation

The checkup shall perform the following checks:

1. Check if Kubevirt is installed and ready.
2. Check if Multus exists.
3. Check if the input NetworkAttachmentDefinition exists.

> 4. There are at least two worker nodes.

Reviewer: I think it should also check if source and/or target nodes the user required exists.
It will probably require nodes viewer permissions.

Author: @ormergi is it possible to limit the data that the checkup will receive to just the Nodes names?

Reviewer: Looking at this again I find it not clear, what I meant is that it should check if source and/or target nodes that were specified exists (not as an additional setp to other checks).

@orelmisan I dont think so, AFAIK RBAC API doesn't provide such granularity (up to object fields).
Note that if the nodes are not exists the checkup will eventually fail when the VMs are created,
having such check will make the checkup fail faster.

Reviewer 2: We have learned that security/permissions is prioritized over fast failure.
@orelmisan , perhaps you can adjust it to be more generic... e.g. `The required nodes are available to run the checkup.`

This should cover implicit and explicit nodes specifications, even futuristic selectors can fall into this spec.

Author: @EdDev I've updated section 4 according to your suggestion.

- Measure latency between two VMI/s, on a predefined network.
- Share results (and logs) with the cluster administrator.

## Non-Goals

- Define or configure the network under test.

# Proposal

## Definition of Users

- Kubernetes Cluster Administrators

## User Stories

As a Kuberenetes Cluster Administrator I would like to:

1. Check the connectivity and latency of a network in my cluster, by:

   1.1. Having an active cluster, with at least two worker nodes.

   1.2. Defining, configuring and creating a NetworkAttachmentDefinition (NAD), which I am interested in checking.

   1.3. Running the Kubevirt Network Latency checkup, while providing as input the NAD.

   1.3.1. Optionally, I want to provide the maximum accepted latency, such that beyond this value, the check will fail.

   1.4. Reviewing the status report (of the run), including:

   1.4.1. Success or failure indication.

   1.4.1.1 Setup failure indication is shown in case of a missing/incorrect input.

   1.4.2 Latency details (min, max, average and period).

1.5. Collecting relevant manifests and logs in case of failure (for troubleshooting).

## API

### Input

> **Note**: The K8s checkup framework provides additional environment variables (for example the output ConfigMap).

Author: Done.

The checkup shall receive its custom configuration from the following environment variables:

| Env Variable Name                | Description                                                                                      | Is Mandatory | Remarks                                                              |
|----------------------------------|--------------------------------------------------------------------------------------------------|--------------|----------------------------------------------------------------------|

> | NAD_NAMESPACE                    | Namespace where the NetworkAttachmentDefinition object resides.                                  | True         |                                                                      |

Reviewer 2: As @phoracek pointed out, we should avoid the use of the name `NAD`. Lets use the full name.

Author: Done.

| NAD_NAME                         | NetworkAttachmentDefinition object name.                                                         | True         |                                                                      |
| SOURCE_NODE                      | Check latency from this node to TARGET_NODE                                                      | False        | When specified the VM will start with a corresponding affinity rule. |
| TARGET_NODE                      | Check latency from SOURCE_NODE to this node                                                      | False        | When specified the VM will start with a corresponding affinity rule. |
| SAMPLE_DURATION_SECONDS          | Latency check duration [sec]                                                                     | False        | Defaults to 5 [sec]                                                  |

> | MAX_DESIRED_LATENCY_MILLISECONDS | Maximum desired latency between VMs. If the latency is higher than this - the checkup fails [ms] | False        | Defaulting to 0. 0 translates to a no-max-latency limit.             |

Reviewer 2: Please try to see if the columns can be shrink a bit (splitting tests into more lines).
